Moving toward making pick_and_place_demo not iiwa specific

sammy-tri opened the pull request:
First step of many.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8357)
<!-- Reviewable:end -->

sammy-tri commented:
+@avalenzu for feature review.  This is mostly just shifting code around and deleting unused bits.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-:-L7WFxoR85zOJs1WcA3Z:blc6bi0)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Ok, I'm starting to pile on a bit now.  I promise no new changes for this grab-bag PR.  Let me know if you want me to split it up, I think each commit can stand alone if we want to spread out the load.

---

Review status: 0 of 19 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-:-L7WLmWB8usEmPNO30Gx:bm1qrbd)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one typo. Given the number of files, you may want to call out to the platform reviewer which ones actually need platform review.

---

Reviewed 6 of 6 files at r1, 4 of 4 files at r2, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-L7zCTAjkNY_OQ64Kbq6:-L7zCTAjkNY_OQ64Kbq7:bmmykl8) ([raw file](https://github.com/robotlocomotion/drake/blob/d143fdaab89a1d4acfb39041875a732fadd6ff0a/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
> ```cpp
>   }
> 
>   if (task.grasp_frame_translational_offset() != 0) {
> ```

BTW s/translational/angular/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-:-L7zDAGFsdCEe2TDcWr3:b-ad8uz1)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@ericcousineau-tri for platform review.  Note that several files are currently in `dev/` (which helpfully sorted to the top 7 of this PR) and do not require platform review.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r3](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zCTAjkNY_OQ64Kbq6:-L7zTx3eApYd0Fxnw_kv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d143fdaab89a1d4acfb39041875a732fadd6ff0a/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW s/translational/angular/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-:-L7zU3V-E0OwImaieedx:b-tj1gsm)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
First pass done, PTAL.
Concerned about brittle duplication and non-generality of the gripper rotation - seems like both of these could be solved with a fixture URDF.
Ended up reviewing some of the `dev/` components, since these issues were also present here.

---

Reviewed 6 of 6 files at r1, 4 of 4 files at r5, 6 of 9 files at r6.
Review status: 16 of 19 files reviewed at latest revision, all discussions resolved.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 28 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zexB49ZO92s49_-IH:-L7zexB49ZO92s49_-II:b4qwjct) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L28)):*
> ```cpp
> // 0.057m thus giving the surface height (`z`) in world coordinates as
> // 0.736 + 0.057 / 2.
> const double kTableTopZInWorld = 0.736 + 0.057 / 2;
> ```

Would it be possible for us to remove this constant math (brittle duplication), and use frames in the SDF?
Perhaps place a frame in the tables, something at the base of one of the legs, and something at the top, and have this infer the table height from that?

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zfQzMCLWIyLnrYn5I:-L7zfQzMCLWIyLnrYn5J:b-wl225p) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
> ```cpp
>       auto wsg_frame = std::make_shared<RigidBodyFrame<double>>(*frame_ee);
>       wsg_frame->get_mutable_transform_to_body()->rotate(
>           Eigen::AngleAxisd(-0.39269908, Eigen::Vector3d::UnitY()));
> ```

Similar to Naveen's request back for the Xtion sensor, can these be constants either (a) be made constants, or (b) - and more preferably - be made into a fixture URDF?

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zg6VU8BfLShmhyaei:-L7zg6VU8BfLShmhyaej:bf7qkg9) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
> ```cpp
>       auto wsg_frame = std::make_shared<RigidBodyFrame<double>>(*frame_ee);
>       wsg_frame->get_mutable_transform_to_body()->rotate(
>           Eigen::AngleAxisd(-0.39269908, Eigen::Vector3d::UnitY()));
> ```

This constant is directly duplicated in the above code for `planner_configuration_.grasp_frame_angular_offset`.
Can that duplication be removed and given semantic meaning?

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 51 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zkdgS9HG7_2EBXj8_:-L7zkdgS9HG7_2EBXj8a:b9psn6l) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L51)):*
> ```objc
>   /// Gripping force to apply (in Newtons).
>   double grip_force{40.};
>   /// The distance from the end effector link to the grasp frame's
> ```

BTW Can I ask if this will be specific to parallel jaw grippers? If so, could that be clarified in this comment?

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 53 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zgRvxCdm_lxvB764v:-L7zgRvxCdm_lxvB764w:bixei29) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L53)):*
> ```objc
>   /// The distance from the end effector link to the grasp frame's
>   /// origin (position between the fingertips of the gripper).
>   double grasp_frame_translational_offset{0.19};
> ```

Having this be a single scalar seems a bit non-general -- it seems to be rotation about the wrist, but I'm not sure if that's stated here.

Is it possible to turn this into a rotation matrix to generalize this?
(Or rather, is there a way that this can be delegated to a fixture URDF for the WSG gripper?)

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 56 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zkJYv4fVphJDaMdce:-L7zjZRr63UcVTiclwQ5:bquf4ju) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L56)):*
> ```protobuf
> 
> // Next ID: 7
> message PickAndPlaceTask {
> ```

BTW Can a comment be made here stating that this will be used when when parsing configs into a `PickAndPlaceConfiguration` structure?

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 70 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zlSqI8rc5CMdnItPr:-L7zlSqI8rc5CMdnItPs:bf4wu6f) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L70)):*
> ```protobuf
>     // The distance from the end effector link to the grasp frame's
>     // origin (position between the fingertips of the gripper).
>     // DEFAULT: 0.19
> ```

From looking at the protobuf docs, it seems that we there is `[default = X]` syntax (for Protobuf2, it seems).

Is there a reason why we do not use that here?

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zhRBE6hUhEcGGKdPg:-L7zhRBE6hUhEcGGKdPh:bn20icx) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
> ```cpp
>   }
> 
>   if (task.grasp_frame_angular_offset() != 0) {
> ```

Is there a reason why we check for `0`, rather than `task.has_{field}()` -- if we won't be using the `[default = X]` syntax?

https://developers.google.com/protocol-buffers/docs/cpptutorial#the-protocol-buffer-api
(I assume this is for ProtoBuf 2 + 3)

---

*[examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration, line 171 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zi8WW5E8IZ77DysQs:-L7zi8WW5E8IZ77DysQt:b-g8nnzh) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration#L171)):*
> ```
>   end_effector_name: "iiwa_link_ee"
>   # -pi / 8
>   grasp_frame_angular_offset: -0.39269908
> ```

Same complaints as above with duplication of the above rotation: Is there any way to delegate this to another configuration component?
Especially since `end_effector_name` is indicated, it seems like either (a) this offset could folded into the ee frame, or (b) this offset could be a secondary frame (if that shifts too many constants / frames downstream).
Will this ever change per task?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-:-L7zikmQ-a4gJgWGAfXs:b-4tlb3m)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 15 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 28 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zexB49ZO92s49_-IH:-L7zrOqM38Zuh9hO_LZT:buuqr49) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L28)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Would it be possible for us to remove this constant math (brittle duplication), and use frames in the SDF?
Perhaps place a frame in the tables, something at the base of one of the legs, and something at the top, and have this infer the table height from that?
</blockquote></details>

Given that this PR doesn't make the issue any worse than it had been when this constant was in lcm_plant.cc before it moved here, does this need to be solved now?  Particularly for dev/ code, I feel like "you touched it now you own it all" is a high bar.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zfQzMCLWIyLnrYn5I:-L7zs6WKAQlCrcfywJad:bpzkqow) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Similar to Naveen's request back for the Xtion sensor, can these be constants either (a) be made constants, or (b) - and more preferably - be made into a fixture URDF?
</blockquote></details>

I suppose this is as good of a place to hash out the "fixture urdf" concept as anywhere...

* Name of urdf containing at least two frames with some spatial relationship between them
* Name of the frame in the fixture urdf to attach to the robot
* Name of the frame in the robot urdf to attach fixture frame above to
* Name of the frame in the fixture urdf which the gripper model should be attached to

This is the general idea?

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zg6VU8BfLShmhyaei:-L7zt7kC1atAiACgbDOx:bv5awp5) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

This constant is directly duplicated in the above code for `planner_configuration_.grasp_frame_angular_offset`.
Can that duplication be removed and given semantic meaning?
</blockquote></details>

One caveat is that we're slightly conflating two different concepts which both involve a -pi/8 rotation, where one is where to place the gripper when creating a simulation (this code here), and the other is where to place the grasp frame relative to a frame in the robot urdf for planning.  In this case the rotational component is the same (though the longitudinal component is not).  Yes, I can move some values out of the task section in the config file and into the robot section if needed, and this would reduce duplication, though the task section (probably) would still need some way to configure the planning-related properties which are not relevant to building the simulation.

I'll wait for the "fixture urdf" discussion to move further along before contemplating how those properties would be specified.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 51 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zkdgS9HG7_2EBXj8_:-L7zwFgx2aq3MzXXhvhk:b-udfxew) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L51)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Can I ask if this will be specific to parallel jaw grippers? If so, could that be clarified in this comment?
</blockquote></details>

Attempted to clarify.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 53 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zgRvxCdm_lxvB764v:-L7zwKHu73_1xBd61ufY:b-kynejv) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L53)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Having this be a single scalar seems a bit non-general -- it seems to be rotation about the wrist, but I'm not sure if that's stated here.

Is it possible to turn this into a rotation matrix to generalize this?
(Or rather, is there a way that this can be delegated to a fixture URDF for the WSG gripper?)
</blockquote></details>

I'll let the fixture urdf discussion play out above and then return to this.  One note is that you'd need a fixture urdf for every possible value of this parameter (part of the point here is to support other grippers, though that's not in this PR).

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 56 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zkJYv4fVphJDaMdce:-L7zwpa_8lnoZQ4qIxgQ:bcup9kx) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Can a comment be made here stating that this will be used when when parsing configs into a `PickAndPlaceConfiguration` structure?
</blockquote></details>

The mapping of protobuf message -> parsed structure is a little more complicated than that.  Let's see where the rest of the discussions about the configuration file wind up and revisit.  If I'm refactoring chunks of the config stuff anyway I'm not sure how this is going to come out.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 70 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zlSqI8rc5CMdnItPr:-L7zxXZx6EMymGhNzsP4:b-xfsmhy) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L70)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

From looking at the protobuf docs, it seems that we there is `[default = X]` syntax (for Protobuf2, it seems).

Is there a reason why we do not use that here?
</blockquote></details>

I don't have any idea, I'm not the original author.  Would you like that refactoring done in this PR, in a PR to be merged before this one and rebase this on top of it, or in a subsequent PR?

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zhRBE6hUhEcGGKdPg:-L7zyFSD386rdp-kPH2x:b9jv9qo) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is there a reason why we check for `0`, rather than `task.has_{field}()` -- if we won't be using the `[default = X]` syntax?

https://developers.google.com/protocol-buffers/docs/cpptutorial#the-protocol-buffer-api
(I assume this is for ProtoBuf 2 + 3)
</blockquote></details>

Fix will depend on the resolution of the "default" issue.

---

*[examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration, line 171 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zi8WW5E8IZ77DysQs:-L7zyMnD0Bh_1KQ_qn69:badamxa) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration#L171)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Same complaints as above with duplication of the above rotation: Is there any way to delegate this to another configuration component?
Especially since `end_effector_name` is indicated, it seems like either (a) this offset could folded into the ee frame, or (b) this offset could be a secondary frame (if that shifts too many constants / frames downstream).
Will this ever change per task?
</blockquote></details>

I suppose it might change per task if we added support for tool changers, but that's a pretty big job at any rate.

Waiting on the other parts of this comment until we hash out the config file / fixture urdf changes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 15 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 28 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zexB49ZO92s49_-IH:-L8--y8P1XvJ8Kh4u5Rr:b-8x8thf) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L28)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Given that this PR doesn't make the issue any worse than it had been when this constant was in lcm_plant.cc before it moved here, does this need to be solved now?  Particularly for dev/ code, I feel like "you touched it now you own it all" is a high bar.
</blockquote></details>

Gotcha. I'd be happy with a TODO in this case; this is just a convention (constants that could be embedded the model) that I'd like to see disappear as soon as possible, given that the code has been alive for quite a while.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zfQzMCLWIyLnrYn5I:-L8-0FP3DCJSjZ6Nkn6S:b-9b179m) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
Yup! With some slight adjustemnts:

> Name of urdf containing at least two frames with some spatial relationship between them

If it's just a fixture URDF, you only need one frame, whose pose can be extracted relative to `base` without having to compute kinematics.  Example:
https://github.com/RobotLocomotion/drake/blob/4ae9651d31246898d4507437a566165ec24c20f4/manipulation/models/xtion_description/urdf/xtion_wsg_fixture.urdf

> Name of the frame in the fixture urdf to attach to the robot

Also not necessary. If you just use `base`, and just add the model instance to the upstream frame:
`https://github.com/RobotLocomotion/drake/blob/65e306d9738949b085258ab9e578eb014ad09f55/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L153-L166`
(not linking, b/c lines are kinda bloat in this comment)

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zg6VU8BfLShmhyaei:-L8-2E5BAcizHOzGuk_e:b-30xsal) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
> One caveat is that we're slightly conflating two different concepts which both involve a -pi/8 rotation, where one is where to place the gripper when creating a simulation (this code here), and the other is where to place the grasp frame relative to a frame in the robot urdf for planning [...]

Er, not sure I exactly understand why they're not related, but will ask via f2f.
(I believe I understand sim vs. reality and plant vs. control / planning, but it seems that we should defer offsets to frames when it comes to physical mounting, for both the plant and control / planning.)

> (though the longitudinal component is not)

Is there a reason why the longitudinal offset has to be different? Is this a disparity between simulation and hardware?

All this being said, I'd be fine if these disparities were fixed in a future PR.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 56 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zkJYv4fVphJDaMdce:-L8-1dPe4RPt-8rzTOfY:b56uezr) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L56)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

The mapping of protobuf message -> parsed structure is a little more complicated than that.  Let's see where the rest of the discussions about the configuration file wind up and revisit.  If I'm refactoring chunks of the config stuff anyway I'm not sure how this is going to come out.
</blockquote></details>

OK Sounds good!

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto, line 70 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zlSqI8rc5CMdnItPr:-L8-1flg6ODr0-fXxlvc:bvlxdb8) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.proto#L70)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I don't have any idea, I'm not the original author.  Would you like that refactoring done in this PR, in a PR to be merged before this one and rebase this on top of it, or in a subsequent PR?
</blockquote></details>

OK I'd be happy with it in a subsequent PR.
If these configuration files are to be extended upon, it would clarify things to have one concrete source of default values.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 15 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zfQzMCLWIyLnrYn5I:-L8-IEECEKI3_LJj-yaH:b-lhaai9) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Yup! With some slight adjustemnts:

> Name of urdf containing at least two frames with some spatial relationship between them

If it's just a fixture URDF, you only need one frame, whose pose can be extracted relative to `base` without having to compute kinematics.  Example:
https://github.com/RobotLocomotion/drake/blob/4ae9651d31246898d4507437a566165ec24c20f4/manipulation/models/xtion_description/urdf/xtion_wsg_fixture.urdf

> Name of the frame in the fixture urdf to attach to the robot

Also not necessary. If you just use `base`, and just add the model instance to the upstream frame:
`https://github.com/RobotLocomotion/drake/blob/65e306d9738949b085258ab9e578eb014ad09f55/drake/examples/kuka_iiwa_arm/dev/monolithic_pick_and_place/monolithic_pick_and_place_demo.cc#L153-L166`
(not linking, b/c lines are kinda bloat in this comment)
</blockquote></details>

That looks like it specifies two frames in that code (`body_frame` on the gripper and `xtion_wsg_fixture` on the fixture)?  Which is still fewer than I originally guessed.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zg6VU8BfLShmhyaei:-L8-Ice3CZjAyhRpF8sK:b-r43wm) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

> One caveat is that we're slightly conflating two different concepts which both involve a -pi/8 rotation, where one is where to place the gripper when creating a simulation (this code here), and the other is where to place the grasp frame relative to a frame in the robot urdf for planning [...]

Er, not sure I exactly understand why they're not related, but will ask via f2f.
(I believe I understand sim vs. reality and plant vs. control / planning, but it seems that we should defer offsets to frames when it comes to physical mounting, for both the plant and control / planning.)

> (though the longitudinal component is not)

Is there a reason why the longitudinal offset has to be different? Is this a disparity between simulation and hardware?

All this being said, I'd be fine if these disparities were fixed in a future PR.
</blockquote></details>

(scribing from f2f) I'll get the patch ready to fix these issues and then either add to this PR or open a new one depending on how large/intrusive it looks. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review status: 15 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-L7zhRBE6hUhEcGGKdPg:-L82ZfHeFbFGy2FdaXsE:b-55zsql) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Fix will depend on the resolution of the "default" issue.
</blockquote></details>

Proto3 has no `has_{field}()` for regular fields (note their absence [here](https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#singular-numeric-fields-proto3)). The only way to get one is to use [Oneof fields](https://developers.google.com/protocol-buffers/docs/proto3#oneof). We could use these for all our scalar fields that have defaults.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8357)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
(from f2f) Dropped the commit which modified the config file, so constant duplication should be no worse than before.  Will open another PR to update how configuration is done eventually.

---

Review status: 8 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 28 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zexB49ZO92s49_-IH:-L88vXXAAqSogeg-yxoV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L28)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Gotcha. I'd be happy with a TODO in this case; this is just a convention (constants that could be embedded the model) that I'd like to see disappear as soon as possible, given that the code has been alive for quite a while.
</blockquote></details>

Done.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zfQzMCLWIyLnrYn5I:-L88vikD6Ft3fFfIaWg1:b-v2flog) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

That looks like it specifies two frames in that code (`body_frame` on the gripper and `xtion_wsg_fixture` on the fixture)?  Which is still fewer than I originally guessed.
</blockquote></details>

From f2f: will fix in future PR.

---

*[examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc, line 73 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zg6VU8BfLShmhyaei:-L88vlsLAJzxVrDzfobD:bpwgs2c) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/dev/pick_and_place/pick_and_place_simulation_helpers.cc#L73)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

(scribing from f2f) I'll get the patch ready to fix these issues and then either add to this PR or open a new one depending on how large/intrusive it looks. 
</blockquote></details>

(again from f2f) Will deal with in future PR.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 53 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zgRvxCdm_lxvB764v:-L88vsoKBxXRwb0nBLtd:b-j0xz8k) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L53)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'll let the fixture urdf discussion play out above and then return to this.  One note is that you'd need a fixture urdf for every possible value of this parameter (part of the point here is to support other grippers, though that's not in this PR).
</blockquote></details>

Will update in future PR.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc, line 199 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zhRBE6hUhEcGGKdPg:-L88vbYXE_zVabjFpPZX:b7my1zi) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration_parsing.cc#L199)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Proto3 has no `has_{field}()` for regular fields (note their absence [here](https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#singular-numeric-fields-proto3)). The only way to get one is to use [Oneof fields](https://developers.google.com/protocol-buffers/docs/proto3#oneof). We could use these for all our scalar fields that have defaults.
</blockquote></details>

Not touching this file in the PR anymore.

---

*[examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration, line 171 at r6](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L7zi8WW5E8IZ77DysQs:-L88vekt6FQ9bpFt7_eN:b-i19xiq) ([raw file](https://github.com/robotlocomotion/drake/blob/ae76bb70b7dbef049f98387d16ccac1771e1d06b/examples/kuka_iiwa_arm/pick_and_place/configuration/dual_iiwa.pick_and_place_configuration#L171)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I suppose it might change per task if we added support for tool changers, but that's a pretty big job at any rate.

Waiting on the other parts of this comment until we hash out the config file / fixture urdf changes.
</blockquote></details>

Removed this file from the PR.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-:-L88ufBPEJ5MaEBRFpju:b-3bnq1c)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  with one BTW.

---

Reviewed 2 of 4 files at r5, 6 of 12 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 51 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-L88wRbqAa8DFaE8B6aO:-L88wRbqAa8DFaE8B6aP:bcujrtp) ([raw file](https://github.com/robotlocomotion/drake/blob/d143fdaab89a1d4acfb39041875a732fadd6ff0a/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L51)):*
> ```objc
>   /// Gripping force to apply (in Newtons).
>   double grip_force{40.};
>   /// The distance from the end effector link to the grasp frame's
> ```

BTW, it looks like we went from `///` to `//` on this description.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8357#-:-L88wlNlX6yP0oqznhNi:bhl8ng0)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h, line 51 at r3](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-L88wRbqAa8DFaE8B6aO:-L88x2gjA2QLN8CFuJ6q:b-ji3jpw) ([raw file](https://github.com/robotlocomotion/drake/blob/d143fdaab89a1d4acfb39041875a732fadd6ff0a/examples/kuka_iiwa_arm/pick_and_place/pick_and_place_configuration.h#L51)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, it looks like we went from `///` to `//` on this description.
</blockquote></details>

True, though I don't think that's a change in this PR.  (or perhaps more accurately this PR had fixed the issue before I removed the commit which modified this file)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8357#-:-L8AFJQc2ewvEv_fy-Ba:bnfp4nl)*
<!-- Sent from Reviewable.io -->
